File: tests/github.py
"""GitHub fixtures and bounded mocked-gh process checks; never contacts an account."""
import importlib.util
import json
import os
import signal
import subprocess
from pathlib import Path
import sys
import tempfile
import time
import unittest
from unittest.mock import patch

source = Path(sys.argv.pop(1)).resolve()
spec = importlib.util.spec_from_file_location("github_status", source)
github = importlib.util.module_from_spec(spec)
spec.loader.exec_module(github)


def pull(number=1, state="SUCCESS"):
    return {"number": number, "title": "<b>Fixture</b>\nchange", "url": f"https://github.com/org/repo/pull/{number}",
            "repository": {"nameWithOwner": "org/repo"}, "reviewDecision": "APPROVED", "isDraft": False,
            "updatedAt": "2026-09-08T00:00:00Z", "commits": {"nodes": [{"commit": {"statusCheckRollup": {"state": state}}}]}}


def fixture():
    return {"data": {"viewer": {"login": "fixture-user", "pullRequests": {"nodes": [pull()], "totalCount": 1}},
                     "search": {"nodes": [pull(2, "FAILURE")], "issueCount": 1}}}


class GitHubTests(unittest.TestCase):
    def test_snapshot_is_two_read_only_requests_with_bounded_query(self):
        calls = []
        def run(args):
            calls.append(args)
            return {"login": "fixture-user"} if len(calls) == 1 else fixture()
        result = github.snapshot(run)
        self.assertEqual(result["state"], "ready")
        self.assertEqual(result["authored"][0]["checks"], "SUCCESS")
        self.assertEqual(result["reviews"][0]["checks"], "FAILURE")
        self.assertNotIn("\n", result["authored"][0]["title"])
        self.assertEqual(len(calls), 2)
        self.assertEqual(calls[0], ["api", "--hostname", "github.com", "user"])
        self.assertIn("reviews=is:pr is:open review-requested:fixture-user sort:updated-desc", calls[1])
        self.assertNotIn("mutation", github.QUERY.lower())
        self.assertIn("first: 20", github.QUERY)

    def test_rows_limit_dedup_draft_and_unknown_checks(self):
        records = [pull(), pull()] + [pull(i) for i in range(2, 50)]
        records[0]["isDraft"] = True
        records[0]["commits"] = {"nodes": [{"commit": {"statusCheckRollup": None}}]}
        entries, total = github.connection({"nodes": records, "totalCount": 80}, "github.com", "totalCount")
        self.assertLessEqual(len(entries), 20)
        self.assertEqual(total, 80)
        self.assertTrue(entries[0]["draft"])
        self.assertEqual(entries[0]["checks"], "UNKNOWN")
        self.assertEqual(len({p["url"] for p in entries}), len(entries))

    def test_urls_are_canonical_https_on_selected_host(self):
        valid = "https://github.com/org/repo/pull/1"
        self.assertEqual(github.safe_url(valid, "github.com"), valid)
        for url in ["javascript:alert(1)", "https://evil.example/org/repo/pull/1", "https://github.com.evil/org/repo/pull/1", "https://user@github.com/org/repo/pull/1", valid + "?token=secret", valid + "#note", "https://github.com/org/repo/issues/1", "https://github.com:443/org/repo/pull/1"]:
            self.assertEqual(github.safe_url(url, "github.com"), "", url)
        self.assertEqual(github.safe_url("https://git.example/org/repo/pull/1", "git.example"), "https://git.example/org/repo/pull/1")

    def test_auth_rate_limit_and_graphql_errors_are_sanitized(self):
        for raw, expected in [("HTTP 401 secret", "auth-required"), ("API rate limit exceeded secret", "rate-limited"), ("connection failed secret", "error")]:
            error = github.failure(raw)
            self.assertEqual(error.state, expected)
            self.assertNotIn("secret", error.message)
        responses = iter([{"login": "fixture-user"}, {"errors": [{"message": "rate limit exceeded SECRET"}]}])
        with self.assertRaises(github.FetchError) as result:
            github.snapshot(lambda _: next(responses))
        self.assertEqual(result.exception.state, "rate-limited")
        self.assertNotIn("SECRET", result.exception.message)

    def test_account_changes_and_invalid_host_fail(self):
        data = fixture(); data["data"]["viewer"]["login"] = "other"
        responses = iter([{"login": "fixture-user"}, data])
        with self.assertRaises(github.FetchError): github.snapshot(lambda _: next(responses))
        for host in ["https://github.com", "github.com/path", "user@github.com", "bad..host"]:
            with self.assertRaises(github.FetchError): github.hostname(host)

    def test_shell_shutdown_kills_and_reaps_cli(self):
        for stop_signal in (signal.SIGTERM, signal.SIGINT):
            with self.subTest(signal=stop_signal), tempfile.TemporaryDirectory() as directory:
                pid_file = Path(directory) / "gh.pid"
                binary = Path(directory) / "gh"
                binary.write_text(f"#!{sys.executable}\nimport os,time\nfrom pathlib import Path\n"
                                  "Path(os.environ['SEELE_GITHUB_TEST_PID']).write_text(str(os.getpid()))\n"
                                  "time.sleep(30)\n")
                binary.chmod(0o755)
                environment = dict(os.environ, PATH=directory, SEELE_GITHUB_TEST_PID=str(pid_file))
                backend = subprocess.Popen([sys.executable, str(source)], env=environment,
                                           stdout=subprocess.DEVNULL, stderr=subprocess.DEVNULL)
                child = None
                try:
                    deadline = time.monotonic() + 3
                    while time.monotonic() < deadline:
                        try:
                            value = pid_file.read_text()
                            if value.isdigit():
                                child = int(value)
                                break
                        except FileNotFoundError:
                            pass
                        time.sleep(0.01)
                    self.assertIsNotNone(child, "mock CLI must have started")
                    backend.send_signal(stop_signal)
                    self.assertEqual(backend.wait(timeout=3), 128 + stop_signal)
                    with self.assertRaises(ProcessLookupError):
                        os.kill(child, 0)
                    child = None
                finally:
                    if backend.poll() is None:
                        backend.kill()
                    backend.wait()
                    if child is not None:
                        try:
                            os.killpg(child, signal.SIGKILL)
                        except ProcessLookupError:
                            pass

    def test_mocked_cli_has_size_timeout_and_noninteractive_bounds(self):
        with tempfile.TemporaryDirectory() as directory:
            binary = Path(directory) / "gh"
            def stub(body):
                binary.write_text(f"#!{sys.executable}\n" + body)
                binary.chmod(0o755)
            with patch.dict(os.environ, {"PATH": directory, "GH_DEBUG": "api"}):
                stub("import os,json\nassert os.environ['GH_PROMPT_DISABLED']=='1'\nassert 'GH_DEBUG' not in os.environ\nprint(json.dumps({'login':'fixture'}))\n")
                self.assertEqual(github.run_gh(["api", "user"]), {"login": "fixture"})
                stub("print('x'*10000)\n")
                with self.assertRaises(github.FetchError) as error: github.run_gh([], max_output=128)
                self.assertIn("more data", error.exception.message)
                stub("import time\ntime.sleep(30)\n")
                start = time.monotonic()
                with self.assertRaises(github.FetchError): github.run_gh([], timeout=0.05)
                self.assertLess(time.monotonic() - start, 2)
                stub("import sys\nprint('HTTP 401 TOKEN',file=sys.stderr)\nsys.exit(1)\n")
                with self.assertRaises(github.FetchError) as error: github.run_gh([])
                self.assertEqual(error.exception.state, "auth-required")
                self.assertNotIn("TOKEN", error.exception.message)


if __name__ == "__main__":
    unittest.main()

File: projects/runtime/src/github.rs
//! Bounded read-only GitHub snapshot. The CLI owns authentication; the UI sees
//! only fixed public errors, never CLI diagnostics or credentials.
use crate::process::{capture, Limits};
use serde_json::{json, Value};
use std::{collections::HashSet, process::Command, sync::atomic::AtomicUsize, time::Duration};

const QUERY: &str = include_str!("github.graphql");
#[derive(Debug)]
pub struct Failure {
    state: &'static str,
    message: &'static str,
}
impl Failure {
    fn error(message: &'static str) -> Self {
        Self {
            state: "error",
            message,
        }
    }
    fn value(&self) -> Value {
        json!({"state": self.state, "message": self.message})
    }
}
fn failure(text: &str) -> Failure {
    let lower = text.to_ascii_lowercase();
    if ["rate limit", "rate_limit", "http 429"]
        .iter()
        .any(|s| lower.contains(s))
    {
        Failure {
            state: "rate-limited",
            message: "GitHub's rate limit was reached. Automatic refresh will wait five minutes.",
        }
    } else if [
        "http 401",
        "bad credentials",
        "authentication",
        "gh auth login",
        "not logged",
    ]
    .iter()
    .any(|s| lower.contains(s))
    {
        Failure {
            state: "auth-required",
            message: "Sign in with GitHub CLI, then refresh.",
        }
    } else {
        Failure::error(
            "GitHub could not be reached. Check your connection and account access, then refresh.",
        )
    }
}
fn hostname(value: &str) -> Result<String, Failure> {
    let host = value.to_ascii_lowercase();
    if host.is_empty()
        || host.len() > 253
        || host.contains("..")
        || !host.as_bytes()[0].is_ascii_alphanumeric()
        || !host.as_bytes()[host.len() - 1].is_ascii_alphanumeric()
        || !host
            .bytes()
            .all(|c| c.is_ascii_alphanumeric() || matches!(c, b'.' | b'-'))
    {
        return Err(Failure::error("The configured GitHub hostname is invalid."));
    }
    Ok(host)
}
/// Canonical pull-request URL grammar shared by source collection and UI intent.
pub fn safe_url<'a>(value: &'a Value, host: &str) -> Option<&'a str> {
    if hostname(host).ok()?.as_str() != host {
        return None;
    }
    let url = value.as_str()?;
    if url.len() > 1024 {
        return None;
    }
    // Literal authority and path reject URL parser normalization discrepancies.
    let path = url
        .strip_prefix("https://")?
        .strip_prefix(host)?
        .strip_prefix('/')?;
    let parts: Vec<_> = path.split('/').collect();
    if parts.len() != 4
        || parts[2] != "pull"
        || parts[3].starts_with('0')
        || parts[3].is_empty()
        || !parts[3].bytes().all(|c| c.is_ascii_digit())
        || !parts[..2].iter().all(|part| {
            !part.is_empty()
                && !matches!(*part, "." | "..")
                && part
                    .bytes()
                    .all(|c| c.is_ascii_alphanumeric() || matches!(c, b'_' | b'-' | b'.'))
        })
    {
        return None;
    }
    Some(url)
}
fn plain(value: &Value, limit: usize) -> String {
    value.as_str().unwrap_or("").chars().take(limit).map(|c| {
        if c.is_control() || matches!(c, '\u{00ad}' | '\u{061c}' | '\u{200b}'..='\u{200f}' | '\u{2028}'..='\u{202e}' | '\u{2060}'..='\u{206f}' | '\u{feff}') { ' ' } else { c }
    }).collect()
}
fn pull(item: &Value, host: &str) -> Option<Value> {
    let url = safe_url(&item["url"], host)?;
    let number = item["number"].as_u64().filter(|number| *number > 0)?;
    let check = item["commits"]["nodes"]
        .as_array()
        .and_then(|nodes| nodes.last())
        .and_then(|node| node["commit"]["statusCheckRollup"]["state"].as_str())
        .unwrap_or("UNKNOWN");
    let checks = if ["SUCCESS", "FAILURE", "PENDING", "ERROR", "EXPECTED"].contains(&check) {
        check
    } else {
        "UNKNOWN"
    };
    let review = item["reviewDecision"].as_str().unwrap_or("");
    let review = if ["APPROVED", "CHANGES_REQUESTED", "REVIEW_REQUIRED"].contains(&review) {
        review
    } else {
        ""
    };
    Some(
        json!({"number": number, "title": plain(&item["title"], 256), "url": url,
        "repository": plain(&item["repository"]["nameWithOwner"], 200), "updatedAt": plain(&item["updatedAt"], 32),
        "draft": item["isDraft"] == true, "checks": checks, "review": review}),
    )
}
fn connection(value: &Value, host: &str, count: &str) -> Result<(Vec<Value>, u64), Failure> {
    let nodes = value["nodes"]
        .as_array()
        .ok_or_else(|| Failure::error("GitHub returned incomplete pull-request data."))?;
    let mut seen = HashSet::new();
    let entries: Vec<_> = nodes
        .iter()
        .take(20)
        .filter_map(|item| pull(item, host))
        .filter(|item| seen.insert(item["url"].as_str().unwrap().to_owned()))
        .collect();
    let total = value[count].as_u64().unwrap_or(0).max(entries.len() as u64);
    Ok((entries, total))
}
fn snapshot(
    mut run: impl FnMut(&[String]) -> Result<Value, Failure>,
    host: &str,
) -> Result<Value, Failure> {
    let host = hostname(host)?;
    let identity = run(&[
        "api".into(),
        "--hostname".into(),
        host.clone(),
        "user".into(),
    ])?;
    let login = identity["login"]
        .as_str()
        .filter(|s| {
            !s.is_empty()
                && s.len() <= 64
                && s.bytes().all(|c| c.is_ascii_alphanumeric() || c == b'-')
        })
        .ok_or_else(|| Failure::error("GitHub did not identify the current account."))?;
    let result = run(&[
        "api".into(),
        "--hostname".into(),
        host.clone(),
        "graphql".into(),
        "-f".into(),
        format!("query={QUERY}"),
        "-f".into(),
        format!("reviews=is:pr is:open review-requested:{login} sort:updated-desc"),
    ])?;
    if let Some(errors) = result["errors"]
        .as_array()
        .filter(|errors| !errors.is_empty())
    {
        return Err(failure(
            &errors
                .iter()
                .filter_map(|item| item["message"].as_str())
                .collect::<Vec<_>>()
                .join(" "),
        ));
    }
    let data = &result["data"];
    if data["viewer"]["login"].as_str() != Some(login) {
        return Err(Failure::error(
            "The GitHub account changed during refresh. Refresh again.",
        ));
    }
    let (authored, authored_total) =
        connection(&data["viewer"]["pullRequests"], &host, "totalCount")?;
    let (reviews, review_total) = connection(&data["search"], &host, "issueCount")?;
    Ok(
        json!({"state": "ready", "message": "", "host": host, "viewer": login,
        "updatedAt": crate::time::timestamp(), "authored": authored, "reviews": reviews,
        "authoredTotal": authored_total, "reviewTotal": review_total}),
    )
}
pub fn run(cancelled: &AtomicUsize) -> Value {
    snapshot(|args| {
        let output = capture(Command::new("gh").args(args).env("GH_PROMPT_DISABLED", "1")
            .env("GH_PAGER", "cat").env("NO_COLOR", "1").env_remove("GH_DEBUG"), b"",
            Limits { timeout: Duration::from_secs(15), output: 256 * 1024 }, cancelled).map_err(|error| {
                use std::io::ErrorKind;
                Failure::error(match error.kind() {
                    ErrorKind::NotFound => "GitHub CLI is unavailable.",
                    ErrorKind::TimedOut => "GitHub took too long to respond. Try refreshing again.",
                    ErrorKind::InvalidData => "GitHub returned more data than this panel can display.",
                    _ => "GitHub could not be reached. Check your connection and account access, then refresh.",
                })
            })?;
        if !output.status.success() { return Err(failure(&String::from_utf8_lossy(&output.stderr))); }
        serde_json::from_slice(&output.stdout).map_err(|_| Failure::error("GitHub returned an unreadable response."))
    }, &std::env::var("SEELE_GITHUB_HOST").unwrap_or_else(|_| "github.com".into())).unwrap_or_else(|error| error.value())
}
#[cfg(test)]
mod tests {
    use super::*;
    #[test]
    fn urls_reject_authority_confusion_and_path_normalization() {
        let valid = json!("https://github.com/org/repo/pull/1");
        assert_eq!(safe_url(&valid, "github.com"), valid.as_str());
        for invalid in [
            "https://github.com.evil/org/repo/pull/1",
            "https://user@github.com/org/repo/pull/1",
            "https://github.com:443/org/repo/pull/1",
            "https://github.com/../repo/pull/1",
            "https://github.com/org/repo/pull/0",
            "https://github.com/org/repo/pull/1?secret",
        ] {
            assert!(
                safe_url(&json!(invalid), "github.com").is_none(),
                "{invalid}"
            );
        }
    }
    #[test]
    fn read_only_snapshot_preserves_the_ui_contract() {
        let row = json!({"number": 1, "url": "https://github.com/org/repo/pull/1", "title": "<b>test</b>\nchange",
            "repository": {"nameWithOwner": "org/repo"}, "isDraft": true,
            "commits": {"nodes": [{"commit": {"statusCheckRollup": {"state": "SUCCESS"}}}]}});
        let mut calls = Vec::new();
        let result = snapshot(|args| {
            calls.push(args.to_vec());
            Ok(if calls.len() == 1 { json!({"login": "fixture"}) } else {
                json!({"data": {"viewer": {"login": "fixture", "pullRequests": {"nodes": [row, row], "totalCount": 3}},
                    "search": {"nodes": [], "issueCount": 0}}})
            })
        }, "github.com").unwrap();
        assert_eq!(calls.len(), 2);
        assert!(!QUERY.contains("mutation"));
        assert_eq!(result["authored"].as_array().unwrap().len(), 1);
        assert_eq!(result["authored"][0]["checks"], "SUCCESS");
        assert_eq!(result["authored"][0]["title"], "<b>test</b> change");
        assert_eq!(result["authoredTotal"], 3);
    }
    #[test]
    fn hostile_shapes_are_rejected_without_panics_or_diagnostics() {
        assert!(pull(
            &json!({"number": 1, "url": "https://github.com/o/r/pull/1", "commits": 3}),
            "github.com"
        )
        .is_some());
        assert!(connection(&json!({"nodes": false}), "github.com", "count").is_err());
        assert_eq!(failure("HTTP 401 SECRET").state, "auth-required");
        assert!(!failure("HTTP 401 SECRET").message.contains("SECRET"));
        for host in ["", "https://github.com", "a..b", "github.com/path"] {
            assert!(hostname(host).is_err());
        }
    }
}
